src/vos/vos_dtx.c

```diff
@@ -717,6 +730,518 @@ vos_dtx_abort_one(struct vos_container *cont, struct daos_tx_id *dti,
+		if (vds->vds_type == DTX_RT_OBJ)
+			rc = vos_dtx_share_obj(umm, dth, dtx, vds, &shared);
+		else
+			rc = vos_dtx_share_key(umm, dth, dtx, vds, &shared);
+		if (rc != 0)
+			return rc;
+
+		d_list_del(&vds->vds_link);
+		D_FREE_PTR(vds);
+	}
+
+	if (rc == 0 && shared)
+		dtx->te_flags |= DTX_EF_SHARES;
+
+	return rc;
+}
+
+/* The caller has started PMDK transaction. */
+void
+vos_dtx_degister_record(struct umem_instance *umm,
+			umem_id_t entry, umem_id_t record, uint32_t type)
+{
+	struct vos_dtx_entry_df		*dtx;
+	struct vos_dtx_record_df	*rec;
+	struct vos_dtx_record_df	*prev = NULL;
+	umem_id_t			 rec_mmid;
+
+	if (!vos_dtx_is_normal_entry(umm, entry))
+		return;
+
+	dtx = umem_id2ptr(umm, entry);
+	rec_mmid = dtx->te_records;
+	while (!UMMID_IS_NULL(rec_mmid)) {
+		rec = umem_id2ptr(umm, rec_mmid);
+		if (!umem_id_equal(umm, record, rec->tr_record)) {
+			prev = rec;
+			rec_mmid = rec->tr_next;
+			continue;
+		}
+
+		if (prev == NULL) {
+			umem_tx_add_ptr(umm, dtx, sizeof(*dtx));
+			dtx->te_records = rec->tr_next;
+		} else {
+			umem_tx_add_ptr(umm, prev, sizeof(*prev));
+			prev->tr_next = rec->tr_next;
+		}
+
+		umem_free(umm, rec_mmid);
+		break;
+	};
+
+	/* The caller will destroy related OBJ/KEY/SVT/EVT record after
+	 * degistered the DTX record. So not reset DTX reference inside
+	 * related OBJ/KEY/SVT/EVT record unless necessary.
+	 */
+	if (type == DTX_RT_OBJ) {
+		struct vos_obj_df	*obj;
+
+		obj = umem_id2ptr(umm, record);
+		if (dtx->te_intent == DAOS_INTENT_UPDATE)
+			obj->vo_dtx_shares--;
+		if (umem_id_equal(umm, obj->vo_dtx, entry) &&
+		    obj->vo_dtx_shares > 0) {
+			umem_tx_add_ptr(umm, obj, sizeof(*obj));
+			obj->vo_dtx = DTX_UMMID_UNKNOWN;
+		}
+	} else if (type == DTX_RT_KEY) {
+		struct vos_krec_df	*key;
+
+		key = umem_id2ptr(umm, record);
+		if (dtx->te_intent == DAOS_INTENT_UPDATE)
+			key->kr_dtx_shares--;
+		if (umem_id_equal(umm, key->kr_dtx, entry) &&
+		    key->kr_dtx_shares > 0) {
+			umem_tx_add_ptr(umm, key, sizeof(*key));
+			key->kr_dtx = DTX_UMMID_UNKNOWN;
+		}
+	}
+}
+
+int
+vos_dtx_prepared(struct daos_tx_handle *dth)
+{
+	struct vos_container	*cont;
+	struct vos_dtx_entry_df	*dtx;
+	int			 rc = 0;
+
+	D_ASSERT(!UMMID_IS_NULL(dth->dth_ent));
+
+	cont = vos_hdl2cont(dth->dth_coh);
+	D_ASSERT(cont != NULL);
+
+	dtx = umem_id2ptr(&cont->vc_pool->vp_umm, dth->dth_ent);
+	if (dth->dth_intent == DAOS_INTENT_UPDATE) {
+		daos_iov_t	kiov;
+		daos_iov_t	riov;
+
+		/* There is CPU yield during the bulk transfer, then it is
+		 * possible that some others (rebuild) abort this DTX by race.
+		 * So we need to locate (or verify) DTX via its ID instead of
+		 * directly using the dth_ent.
+		 */
+		daos_iov_set(&kiov, &dth->dth_xid, sizeof(struct daos_tx_id));
+		daos_iov_set(&riov, NULL, 0);
+		rc = dbtree_lookup(cont->vc_dtx_active_hdl, &kiov, &riov);
+		if (rc == -DER_NONEXIST)
+			/* The DTX has been aborted by race, notify the RPC
+			 * sponsor to retry via returning -DER_INPROGRESS.
+			 */
```

Comment:
I don't quite see why EINPROGRESS is returned in such case, shouldn't we return a different error code since retry won't help in such case?

Reply:
If the client retry, the related modification will be re-executed, that will overwrite the former abort.

Reply:
I see, I thought it's to inform leader to retry global commit.